src/observeProps.js

```diff
@@ -17,52 +27,53 @@ export default function watchProps(TargetComponent) {
     startWatch = () => {
       const protoKeys = Object.getOwnPropertyNames(TargetComponent.prototype);
 
-      this.filterListeners(protoKeys)
-        .subscribe(
-          (listener) => {
+      from(this.filterListeners(protoKeys))
+        .subscribe({
+          next: (listener) => {
             const propSelector = createSelectorForFunction(listener);
 
-            this.propsChange$
-              .map(propSelector)
-              .distinctUntilChanged()
-              .subscribe(this.createObserver(propSelector, listener));
+            this.propsChange$.pipe(
+              distinctUntilChanged((p, q) => propSelector(p.next) === propSelector(q.next)),
+              map(({ next, prev }) => ({ next: propSelector(next), prev: propSelector(prev) })),
+            )
+            .subscribe(this.createObserver(listener));
           },
-          throwError,
-        );
+          error: throwError,
+        });
     };
 
     /**
-     * Create an observable which emits only "prop listener" methods
+     * Filter and emit only "prop listener" methods
      * @param {string[]} protoKeys
-     * @return {Observable<Function>}
+     * @return {Function[]}
      * */
     filterListeners = protoKeys =>
-      Observable
-        .from(protoKeys)
+      protoKeys
         .map(name => TargetComponent.prototype[name])
         .filter(method => method[IS_PROP_LISTENER]);
 
     /**
-     * Create an observable which emits only "prop listener" methods
+     * Create an observer which calls PROP_LISTENER function with next and prev props
      * @param {Function} propSelector
      * @param {Function} fn -
      * @return {Observable<Function>}
      * */
-    createObserver = (propSelector, fn) =>
-      Observer.create(
-        (next) => {
-          const prev = propSelector(this.props);
-          fn.call(this, next, prev);
-        },
-        throwError,
-      );
+    createObserver = fn =>
+      ({
+        next: ({ next, prev }) =>
+          fn.call(this, next, prev),
+        error: throwError,
+      });
 
-    componentWillReceiveProps(nextProps) {
-      this.propsChange$.onNext(nextProps);
+    componentDidUpdate(prevProps) {
+      this.propsChange$.next({
+        next: this.props,
+        prev: { ...prevProps }
```

Comment:
`prev: prevProps` no need to spread

Reply:
@McReader, I think I've just realised that my spread operator and a will not to use a reference to the existent object causes more **redundant memory usage** and **garbage collector work** which really does not make any efficiency here, am I right?